x/emissions/keeper/inference_synthesis/forecast_implied.go

```diff
@@ -0,0 +1,182 @@
+package inferencesynthesis
+
+import (
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	"cosmossdk.io/log"
+	alloraMath "github.com/allora-network/allora-chain/math"
+	emissionstypes "github.com/allora-network/allora-chain/x/emissions/types"
+)
+
+// Calculate the forecast-implied inferences I_ik given inferences, forecasts and network losses.
+// Calculates R_ijk, w_ijk, and I_ik for each forecast k and forecast element (forcast of worker loss) j
+//
+// Forecast without inference => weight in calculation of I_ik and I_i set to 0. Use latest available regret R_i-1,l
+// Inference without forecast => only weight in calculation of I_ik set to 0
+// A value of 0 => no inference corresponded to any of the forecasts from a forecaster
+func CalcForecastImpliedInferences(
+	logger log.Logger,
+	topicId uint64,
+	allInferersAreNew bool,
+	inferers []Inferer,
+	infererToInference map[Inferer]*emissionstypes.Inference,
+	infererToRegret map[Inferer]*Regret,
+	forecasters []Forecaster,
+	forecasterToForecast map[Forecaster]*emissionstypes.Forecast,
+	forecasterToRegret map[Forecaster]*Regret,
+	networkCombinedLoss alloraMath.Dec,
+	epsilonTopic alloraMath.Dec,
+	pNorm alloraMath.Dec,
+	cNorm alloraMath.Dec,
+) (
```

Comment:
so I disagree with struct parameters and the reason why is because one of the linters, will pick up if an input is not being used and flag it for compile / fail CI. Even with exhaustruct on which is coming in my next PR, you won't get that if you start passing a struct which is basically a context struct. If at a later date somebody comes along and modifies this function to not need or use one of its parameters, the compiler will force them to update the function definition and adjust everywhere it's called.

Is this way uglier, yes, is it significantly uglier or harder to read, not really, just more scrolling down when you scroll through code. This is a more declarative approach. And escpecially with this important, math heavy inference synthesis code, I would rather be more declarative and bonedead obvious in what's going on

Reply:
put it another way


nobody has ever gotten burned from having too many args to a function in their code

(actually this is not true in languages like C iirc, but the golang compiler can behind my back stuff this into a struct under the hood, shh)

Reply:
having a long list of arguments still increases the cognitive load for developers, especially for future maintenance. It’s harder to immediately understand what’s happening when the function signature is bloated imo, and it becomes more error-prone when passing the wrong values in large lists of parameters.

Is it impossible to get linter coverage in context struct? If yes, why not at least group static parameters into a logical struct? It would enhance maintainability in the long term and reduce refactor scope when adding or removing fields​.

I understand the desire to keep things "bone-dead obvious," but declarative doesn’t have to mean verbose. Grouping parameters into domain-specific structs still lets the code be clear while reducing the noise of many unrelated arguments. Being more declarative can also mean being more concise, and it's easier to reason about logical groups of data, especially in math heavy code​​.

While I agree that we should keep things declarative and obvious, I think organizing parameters into structs still helps maintain the code’s clarity and readability. Instead of scrolling through a large list of arguments, we can reason about the functionality more easily by working with grouped, coherent structs. The clarity here comes from proper modeling rather than simply expanding the number of args​​.

Also, go compiler might optimize large argument lists under the hood by "stuffing them into a struct", but this does not solve the core problem of code readability, clarity, and maintainability, which are human-facing issues.

Reply:
yeah I don't agree with you and I don't view the readability impact as that bad, even in this explicit example CalcForecastImpliedInferences, the only kind of "tertiary" not directly used parameter is the logger. Everything else is pretty core and used directly in the function, and I wouldn't want it obscured in some struct

Reply:
@guilherme-brandao 
>  think it would be a good idea to have a shared struct with static properties only (e.g., logger, topicId, networkLoss, moduleParams, etc.).
Good thought in theory, let's keep going

@guilherme-brandao 
> It would avoid functions with this huge number of args
If we're at this point (imho probably so in this case), then the entire function signature needs to be a struct, and we need to enforce using named struct fields and `exhauststruct`.  It needs to be a stricter version of Python `kwargs` if it's going to provide guaranteed value to us on the team

@relyt29 
> one of the linters, will pick up if an input is not being used and flag it for compile / fail CI
What do you mean- which linter?  Opinion on keeping it enabled?  That sounds like a stupid fucking linter, but maybe I'm not understanding it fully.  Sometimes you have to pass structs into functions and not every member is used.  Am I misunderstanding?  Can we make it less stupid?  And what is its value proposition?  Unused args are reasonable only in the case of conforming to interfaces, but this sounds more rigid?  Sry lack the full picture

@guilherme-brandao 
> but declarative doesn’t have to mean verbose
Yeah, but explicit and verbose are not fully overlapping, and explicit is a must.  Go's culture leans towards explicit, and orthogonally, it also often leans towards verbose.  My advice: be explicit always, whether with Go or any other language.  Fuck magic.  Be verbose when being explicit requires it.  If we dislike verbosity, we can make simple IDE tools to hide it (I have made tools to hide `if err != nil` blocks when I press a simple keystroke -- in other words, let's think creatively about problems like these, don't change code because of our friction with it, let's change our view of code instead -- metamagic)

@guilherme-brandao 
> Grouping parameters into domain-specific structs still lets the code be clear
Agree.  This is good for human habits of mind.  If a linter dislikes it, disembowel the linter. Lose it or rewrite it.  Human cognitive sympathy with code is very important, especially on complex projects

@guilherme-brandao 
> Also, go compiler might optimize large argument lists under the hood by "stuffing them into a struct"
This is the least of our worries, tbh.  If you think it has become a problem, run `pprof` and prove it, and then we fix it.  No premature optimization, only data

Summary of my opinion:
- Be as explicit as possible, always
- Collapsing a large argument list into a struct (`MyFuncArgs{}` etc) is almost always good once the arg list gets too big, but we need to enforce 1) exhauststruct and 2) requiring field names on struct literals.  Think Python `kwargs` with `exhaust` rules
- Grouping related subdomains of a struct when it makes sense for human cognition is a win for mental energy
- However, if you're going to pass a pre-defined struct full of defaults into one of these subdomain fields, a lot of care should be taken (again, think about human cognition, and ways people might accidentally overlook the details of the interface you are creating by doing this).  This has to be a very well-named and well-defined default struct val.  For me, this is very much a case-by-case assessment, and I can't give many global rules
- Don't worry about struct-packing for performance improvements right now.  We have much bigger performance issues to solve; this will not move the needle on that at all

Reply:
My thoughts are in the description of this PR, which presents and alternative path for the intention of this PR:
https://github.com/allora-network/allora-chain/pull/621

Reply:
resolution: Every function gets its own input struct

Reply:
resolved in [48c996d](https://github.com/allora-network/allora-chain/pull/538/commits/48c996d646c0d933b04b8be3e3fdf68c77b65a57)